autopush/router/apnsrouter.py

```diff
@@ -27,78 +27,129 @@ class APNSRouter(object):
-        self.default_title = router_conf.get("default_title", "SimplePush")
-        self.default_body = router_conf.get("default_body", "New Alert")
-        self._connect()
+        self.apns = dict()
+        self._config = router_conf
+        for channel in self._config.keys():
+            self.apns[channel] = self._connect(self._config[channel])
+        self._connect(router_conf)
         self.log.debug("Starting APNS router...")
         self.ap_settings = ap_settings
 
-    def register(self, uaid, router_data, *args, **kwargs):
-        """Validate that an APNs instance token is in the ``router_data``"""
+    def register(self, uaid, router_data, router_token="firefox",
+                 *args, **kwargs):
+        """Register an endpoint for APNS, on the `router_token` channel.
+
+        This will validate that an APNs instance token is in the
+        ``router_data``,
+
+        :param uaid: User Agent Identifier
+        :type uaid: str
+        :param router_data: Dict containing router specific configuration info
+        :type router_data: dict
+        :param router_token: The channel identifier for cert info lookup
+        :type router_token: str
+
+        returns a modified router_data dict to be stored
+        in the user agent record.
+
+        """
+        if router_token not in self.apns:
+            raise RouterException("Unknown channel specified",
+                                  status_code=400,
+                                  response_body="Unknown channel")
         if not router_data.get("token"):
             raise RouterException("No token registered", status_code=500,
                                   response_body="No token registered")
+        router_data["channel"] = router_token
         return router_data
 
     def amend_msg(self, msg, router_data=None):
+        """This function is stubbed out for this router"""
         return msg
 
     def route_notification(self, notification, uaid_data):
-        """Start the APNS notification routing, returns a deferred"""
+        """Start the APNS notification routing, returns a deferred
+
+        :param notification: Notification data to send
+        :type notification: dict
+        :param uaid_data: User Agent specific data
+        :type uaid_data: dict
+
+        """
         router_data = uaid_data["router_data"]
         # Kick the entire notification routing off to a thread
         return deferToThread(self._route, notification, router_data)
 
     def _route(self, notification, router_data):
-        """Blocking APNS call to route the notification"""
-        token = router_data["token"]
+        """Blocking APNS call to route the notification
+
+        :param notification: Notification data to send
+        :type notification: dict
+        :param router_data: Pre-initialized data for this connection
+        :type router_data: dict
+
+        """
+        router_token = router_data["token"]
+        self._channel = router_data["channel"]
```

Comment:
ISTM there's only [one instance](https://github.com/mozilla-services/autopush/blob/master/autopush/settings.py#L200) of `APNSRouter` per app, so this will clobber `self._channel` for concurrent requests (if one request comes in, then another one comes in while we're still processing the first one).